### PR TITLE
[CONTINT-3776] Fix dual shipping recovery in logs client 

### PR DESCRIPTION
**Pull request by AliDatadog**
### What does this PR do?
We noticed that the container image and SBOM e2e tests were flaky because the images were not received by the fakeintake when we expected them to be.

We found that the image was sent to the fakeintake but the following log was spammed every two minutes:
```
2024-03-14 17:05:02 UTC | CORE | WARN | (pkg/logs/client/http/destination.go:230 in sendAndRetry) | http://fakeintake-service:80/api/v2/sbom: sleeping until 2024-03-14 17:07:02.411400667 +0000 UTC m=+143.352370720 before retrying. Backoff duration 2m0s due to 7 errors
```
It turns out that the fakeintake started after the agent, so for the first minutes we received only HTTP 503 responses.
As you can see in this [notebook](https://dddev.datadoghq.com/notebook/7671321/investigation-of-testsbom-and-testcontainerimage-flakiness), the fakeintake would still receive a payload every 2 minutes.
However, we still had the log because the number of errors was never decreased.

After deep investigation, it turns out that the `retryLock` seems to be responsible here. Before the number of errors could get updated, another goroutine would lock it.

Finally, I believe we should get a fast recovery by default.

### Describe how to test/QA your changes
Deploy the agent on kubernetes with the fakeintake, sbom collection and dual shipping enabled. 

For sbom collection with helm, set the following to `true`: https://github.com/DataDog/helm-charts/blob/40d8e3313397b3f901f0b65af81c2be04add0fe6/charts/datadog/values.yaml#L707-L720

For the fakeintake, add the following env vars to your agent:
```
    - name: DD_CONTAINER_IMAGE_ADDITIONAL_ENDPOINTS
      value: '[{"host": "fakeintake-service:80", "use_ssl": false}]'
    - name: DD_SBOM_ADDITIONAL_ENDPOINTS
      value: '[{"host": "fakeintake-service:80", "use_ssl": false }]'
    - name: DD_SBOM_SENDER_RECOVERY_RESET
      value: 'true'
    - name: DD_ADDITIONAL_ENDPOINTS
      value: '[{"fakeintake-service:80": ["FAKEAPIKEY"], "use_ssl": false }]'
```

Finally, deploy the fakeintake with 
```
--- 
apiVersion: apps/v1
kind: Deployment
metadata:
  name: fakeintake-deployment
spec:
  replicas: 1
  selector:
    matchLabels:
      app: fakeintake
  template:
    metadata:
      labels:
        app: fakeintake
      annotations:
        ad.datadoghq.com/fakeintake.checks: |-
          {
            "openmetrics": {
              "init_config": {},
              "instances": [
                {
                  "openmetrics_endpoint": "http://%%host%%/metrics",
                  "namespace": "fakeintake",
                  "metrics": [".*"]
                }
              ]
            }
          }
    spec:
      containers:
      - name: fakeintake
        image: public.ecr.aws/datadog/fakeintake:latest
        ports:
        - containerPort: 80
---
apiVersion: v1
kind: Service
metadata:
  name: fakeintake-service
spec:
  selector:
    app: fakeintake
  ports:
    - protocol: TCP
      port: 80
      targetPort: 80
```

In order to reproduce, deploy the agent first then deploy the fakeintake after the above log is shown (approximatively 3 minutes after agent starts).

Deploy the fakeintake, the log should not appear anymore and the fakeintake should receive a lot more payloads.

### CODE REVIEW

File: pkg/logs/client/http/destination.go [hunk 1]
```diff
@@ -223,15 +223,17 @@ func (d *Destination) sendAndRetry(payload *message.Payload, output chan *messag
 
 		d.retryLock.Lock()
 		backoffDuration := d.backoff.GetBackoffDuration(d.nbErrors)
+		d.retryLock.Unlock()
 		d.blockedUntil = time.Now().Add(backoffDuration)
 		if d.blockedUntil.After(time.Now()) {
```
Comment (AliDatadog) [inline pkg/logs/client/http/destination.go:228]: Avoid locking the mutex for 2 minutes

File: pkg/config/setup/config.go [hunk 2]
```diff
@@ -2035,7 +2035,7 @@ func bindEnvAndSetLogsConfigKeys(config pkgconfigmodel.Config, prefix string) {
 	config.BindEnvAndSetDefault(prefix+"sender_backoff_base", DefaultLogsSenderBackoffBase)
 	config.BindEnvAndSetDefault(prefix+"sender_backoff_max", DefaultLogsSenderBackoffMax)
 	config.BindEnvAndSetDefault(prefix+"sender_recovery_interval", DefaultForwarderRecoveryInterval)
-	config.BindEnvAndSetDefault(prefix+"sender_recovery_reset", false)
+	config.BindEnvAndSetDefault(prefix+"sender_recovery_reset", true)
```
Comment (gh123man) [inline pkg/config/setup/config.go:2038]: Was changing this default intentional?

**Review by gh123man — APPROVED** [review]

File: pkg/config/setup/config.go (hunk 2 above)
Comment (AliDatadog) [reply pkg/config/setup/config.go:2038]: yes I think it should be normal

**Review by ogaca-dd — APPROVED** [review]

**Review by ogaca-dd — APPROVED** [review]
LGTM for files owned by Agent shared components.

File: pkg/config/setup/config.go (hunk 2 above)
Comment (vickenty) [reply pkg/config/setup/config.go:2038]: We should not change this without discussing with intake. The slow recovery is a feature, not a bug.

**Review by vickenty — CHANGES_REQUESTED** [review]

File: pkg/logs/client/http/destination.go [hunk 3]
```diff
@@ -223,15 +223,17 @@ func (d *Destination) sendAndRetry(payload *message.Payload, output chan *messag
...
+			d.retryLock.Lock()
 			log.Warnf("%s: sleeping until %v before retrying. Backoff duration %s due to %d errors", d.url, d.blockedUntil, backoffDuration.String(), d.nbErrors)
```
Comment (vickenty) [inline pkg/logs/client/http/destination.go:230]: Will be printed on every concurrent send when the EP is not available? Previously only the first send entering sleep would print this (the rest waiting on a lock).

File: pkg/logs/client/http/destination.go [hunk 4]
```diff
@@ -223,15 +223,17 @@ func (d *Destination) sendAndRetry(payload *message.Payload, output chan *messag
 
 		d.retryLock.Lock()
 		backoffDuration := d.backoff.GetBackoffDuration(d.nbErrors)
+		d.retryLock.Unlock()
 		d.blockedUntil = time.Now().Add(backoffDuration)
```
Comment (vickenty) [inline pkg/logs/client/http/destination.go:227]: This is now a data race.

File: pkg/config/setup/config.go (hunk 2 above)
Comment (AliDatadog) [reply pkg/config/setup/config.go:2038]: Alright, I'll revert this change then

File: pkg/logs/client/http/destination.go (hunk 3 above)
Comment (AliDatadog) [reply pkg/logs/client/http/destination.go:230]: Indeed but in the past we had the first send that would not reset the counter after success

File: pkg/logs/client/http/destination.go (hunk 4 above)
Comment (AliDatadog) [reply pkg/logs/client/http/destination.go:227]: Could you expand on that ? How does the behavior change ?

File: releasenotes/notes/fix-dual-shipping-9bab40a3a0f216f6.yaml [hunk 5]
```diff
@@ -0,0 +1,5 @@
+---
+fixes:
+  - |
+    Fixed a bug in the dual shipping feature where events were not being
```
Comment (alai97) [inline releasenotes/notes/fix-dual-shipping-9bab40a3a0f216f6.yaml:4]:
```suggestion
    Fixed a bug in the Dual Shipping feature where events were not being
```

**Review by alai97 — APPROVED** [review]
Looks good for docs!

File: pkg/logs/client/http/destination.go (hunk 4 above)
Comment (vickenty) [reply pkg/logs/client/http/destination.go:228]: I mean the assignment to `blockedUntil`: two threads writing to the same location without synchronization, a definition for a write-write data race. We may end up with weird value inside the `blockedUntil`. Also later, in `waitForBackoff` we read it without synchronization too, which is a read-write race. We may read a value from `blockedUntil` that we didn't write.

Comment (AliDatadog) [reply pkg/logs/client/http/destination.go]: Indeed, sorry for missing that. I suppose we could make `blockedUntil` a local variable then which I did in the last commit. Do you have any other suggestion ?

**Review by L3n41c — APPROVED** [review]

**Review by vickenty — APPROVED** [review]

**Pull request merged**